Automattic/wp-calypso
# Framework: persist redux state

## Persisting Redux State
As part of the offline investigation made in our Miami Calypso Core Meetup, this PR persists redux state using localforage, and loads it as the initial state on boot. This may provide us with some minor performance improvements, and compliments our offline use case. We currently clear this state when users log off.

### ~~Blockers~~
~~( #2757, #2759 ) the redux tree is currently not "pure". Some sections of the tree contain functions, which means that errors are thrown during serialization. Unless we wish to write custom serializers/deserializers, the redux tree should be cleaned up to ensure that it only holds simple objects that do not contain blobs or functions.~~

Update: I'm currently working around this by adding in `SERIALIZE` and `DESERIALIZE` actions that are triggered during loading/before writing to localforage. These actions are not dispatched normally and will not modify the existing store. It instead modifies the copy of current state before we write to localforage, and the initialState object we use when creating a store respectively.

In each reducer, `SERIALIZE` takes what's in memory and returns a simple JS object. Similarly in `DESERIALIZE` the reducer is given state that is a simple JS object. It should validate if the data is valid using a schema. If it is valid, further manipulate the JS object to match what your subtree expects in memory (this might be a no-op, or say instantiation of Immutable.js objects). If the data is invalid, return the default initial state.

### Open questions ###
* How do we deal with loading stale or malformed state? **TBD**
* Should we allow sections of the tree to use immutable state? ( We'd need to create immutable JS instances for those sections ). **Yes, this is allowed provided that the sub-tree handles SERIALIZE and DESERIALIZE correctly**
* Should we allow cycles to exist in our redux state? **No**
* What common conventions should be enforced between each sub-tree? **TBD**

### Action Items ###
- [x] In an effort to limit the size of this PR, I will return default initial state for both `SERIALIZE` and `DESERIALIZE` for many sub-trees. I will create an issue for each piece of the subtree to implement persistence.
- [x] Create issue to move over calls to localStorage and store over to localForage. (Do we run into any specific storage issues if those other keys are on indexDB vs localStorage?) ( #2972 )
- [x] Lots more tests for bad data cases
 
### Subtree Status ###
- [x] state/application (connection status) - Never Persisted 
- [x] state/current-user - Persisted
- [x] state/notices - Never Persisted
- [ ] state/plugins - (Safe to merge). Persistence needs to be implemented ( #2967 )
- [ ] state/posts - (Safe to merge). Persistence needs to be implemented ( #2968 )
- [x] state/sharing/publicize - Persisted
- [ ] state/site-settings/exporter - (Safe to merge). Persistence needs to be implemented. ( #2969 )
- [ ] state/sites/plans - (Safe to merge) Persistence needs to be implemented ( #2970 )
- [x] state/sites - Persisted
- [x] state/support - Never Persisted
- [x] state/themes - Persisted
- [ ] state/ui - (Safe to merge). Persistence needs to be implemented. ( #2971 )
- [x] state/users - Persisted

### Testing instructions ###

* There shouldn't be any visible changes to the site, and we should be able to walk through our normal flows without errors. You should see new debug messages like the following each time a page is loaded:
<img width="961" alt="debug" src="https://cloud.githubusercontent.com/assets/1270189/12763088/633d675e-c9a7-11e5-8fe8-a4ec8e4f4ce0.png">
* After a user logs out, the 'redux-state' should be cleared from indexDB.

Persisted sub-trees to test, with pings for folks who have the most changes in the reducers:
* state/themes: http://calypso.localhost:3000/design works as expected ( @ehg @ockham )
* state/current-user and state/users: loads OK on page load ( @aduth )
* state/sharing/publicize: Start a post that has a publicized connection, reload the editor page, persisted data should load without validation errors. ( @aduth ) 
* state/sites: Navigate to 'My Sites', and switch to a few different sites. Attempt to reload the page. Persisted data should load without validation errors. ( @drewblaisdell )




cc @rralian @mtias @retrofox @artpi @aduth @dmsnell
